### ISSUE TITLE
wxPython 3.0 Support

### ISSUE THREAD

**Issue by Ortham** [wrye-bash/wrye-bash#15]
@Utumno, @lojack5, @D4id4los: I was just on the wxPython site, and I noticed that they don't have their 2.8 releases listed any more. While the binaries are still available on Sourceforge and through the Wrye Python packages, this might be a good indication that we should move to wxPython 3.0. Is this a goal?

The [changelog](http://wxpython.org/recentchanges.php) looks fairly detailed, and so it probably wouldn't be too much work. I'm not familiar with wxPython, but I thought wxWidgets 3.0 was a massive improvement over wxWidgets 2.8 (which I plain refused to use).

While IMHO the ideal goal would be to move to Python 3.x and wxPython's Project Phoenix, AFAIK the latter isn't ready for stable use, but Project Phoenix will still be based on wxWidgets 3.0, so should we ever move to it, this would still be a good first step.

**Comment by Utumno** [comment]
Back burner - notice though that it is taken care of at #163. The de-wx'ing of the code is an essential step towards being able to swap our graphics library without rewriting the UI

**Comment by Metallicow** [comment]
@wrye-bash/bashers 
My metallicow-wx3-phoenix-compatability branch now works on wx28 Classic, wx30 Classic and Phoenix!!! :)
Those interested for the fixes and testing can check it out now!

Notes on my wx3-phoenix compat branch @ runtime currently: 
* wx302 Classic is Crashing hard on wxEVT_CLOSE which is very strange considering wx28 and Phoenix has no problems closing. There is a note in the def that says exceptions have been swallowed for like forever, so this is apparent the problem is with bash doing something funny during startup, runtime, or at closing, which wx3 Classic catches differently or may be deprecated/removed. 
* When running the code on Phoenix, there is still many small fixes to work on: Ex: wizards, mixins, etc

But at least I can say now __OFFICIALLY__ Wrye Bash works on wx3.0.2.0 Classic and Phoenix snapshots, even if poorly because of bash's refactoring at the moment. This is major Step forward if nothing else.

**Comment by Utumno** [comment]
> even if poorly because of bash's refactoring at the moment. 

?

**Comment by Metallicow** [comment]
There are probably some other wx fixes to be found and an Crash(No trace/swallowed) on closing issue lost somewhere in the code that needs fixed.

**Comment by Utumno** [comment]
Reopening this - although the chit chat above is not really fir for a technical issue like this.

Crash is fixed in my local branches - will push soon - due to (surpise !) the poorly coded Ini tabs UI.

Although this is still not an immediate priority it is an interesting programming problem and a blocker for #243 - I need webview to bin the wx.lib.iewin piece of trash and the comtypes dependency so it's high time the eye is turned on this.

Blocked by #163, #190

**Comment by Utumno** [comment]
Ok three things remain:

- [ ] wxLocale crash - see : https://afkmods.iguanadons.net/index.php?/topic/4966-wrye-bash-all-games/&do=findComment&comment=167933 and https://github.com/wrye-bash/wrye-bash/commit/ab9c4a99070a2959f3b209ed16604911b4f895d6
- [ ] cancel does not work on installers tab refresh
- [ ] setCancel does not work on progress: https://stackoverflow.com/questions/42890766/wx-python-3-0-2-classic-disable-the-cancel-button-of-wx-progressdialog?noredirect=1#comment73055118_42890766

@nycz @Metallicow roll up your sleeves :)

**Comment by Metallicow** [comment]
* Well, I'm not sure about the progress dialogs cancel button(It has been known to not work sometimes on various platforms) other than writing a python implementation or try Andrea's pure-python agw [PyProgress](https://github.com/wxWidgets/Phoenix/blob/master/wx/lib/agw/pyprogress.py) . It is probably a wxWidgets issue...
* As far as the wxLocale, I don't have too much experience dealing with it or translations outside of my SourceCoder, so it looks to me like it was introduced somewhere in your code or with wxPy3.
As far as support you might ask on the [dev mailing list](https://groups.google.com/forum/#!forum/wxpython-dev) or [users mailing list](https://groups.google.com/forum/#!forum/wxpython-users)

Also note "Phoenix" is now in beta on PyPI(Close to a official RELz), so "Classic" development has stopped according to Robin(unless you want to maintain it yourself). Try running it with Phoenix to see if the problems still persist or resort to the mailing lists and good-old-fashioned debugging. If you have made it this far, Phoenix is just a commit away.

Oh, and I finally released my widgets package, if you are still looking for that ThreeWaySplitter.
https://github.com/Metallicow/MCOW
The package is for phoenix, tho some work on classic out of the box, tho I have classic files also if someone needs them.

**Comment by nycz** [comment]
@Utumno agh, so sorry for the massively slow reply! (and the general inactivity on my part)
my work irl has been hectic and will most likely continue to keep me busy at least for the next month or two. i still get updates and notifications and stuff but without any real time to do any code here i haven't felt it all that useful to chime in... so yeah, sorry about the radio silence :/

**Comment by Utumno** [comment]
Hey @nycz - glad you are still around :) will keep bugging you - when you get round to it do rebase your branches on current dev - some of it is already in there. I really want at least your work on frames and iewin in, that's very important - and hopefully merge in linux compat changes. Good luck with your job, get rich soon so you can be in here full time :D

**Comment by Utumno** [comment]
And looking at wx.Locale - that's really the biggest hurdle to wx3

**Comment by fireundubh** [comment]
I've been working on locale stuff.

- The locale crash on launch is fixed.
- Translations use canonical names.
- Invalid user-specified languages in `--Language` are overridden to the system default or `en_US` as a fallback.

Right now, I'm just trying to solve why selecting languages from the Languages submenu doesn't pop up a dialog on Windows 10.

```python
# settings_links.py:
# class Settings_Language(RadioLink)
    def Execute(self):
        if self._lang == _bassLang(): return
        if balt.askYes(Link.Frame,
                _(u'Wrye Bash needs to restart to change languages.  Do you '
                  u'want to restart?'), _(u'Restart Wrye Bash')):
            Link.Frame.Restart(['--Language', self._lang])
```

If I set `balt.canVista` to `False`, I get a dialog and the app restarts correctly.

Otherwise, `balt.askYes` immediately returns `False` without showing the dialog and without restarting the app.

Do we really care about showing a styled dialog for changing languages?

@Utumno @lojack5

**Comment by Utumno** [comment]
Great @fireundubh - feel free to push code on a branch here so people can have a look

No we (I) don't - this may point to problems with this vista hack however, that may need be solved. As Robin Dunn stated here: https://stackoverflow.com/a/38017173/281545 windows in wx3 became native  - not sure what this means but probably it means they are actual windows API (I have the feeling that BAIN refresh when many projects are present got faster because of that, we do want native IIUC)

Anyway this vista thing needs revisiting

**Comment by fireundubh** [comment]
Okay. I set `balt.canVista` to `False` in my local branch, which happily makes a number of other dialogs work, too.

I was looking into persisting language settings. Is there any reason why we're using RegEx to parse INI files instead of the native `configparser` library?

**Comment by fireundubh** [comment]
> Setting this to false will affect everything - can you debug and see what call fails?

I don't know what's failing because I don't know what's supposed to be returned.

Best guess:
- `balt.vistaDialog()` returns `(None, checkbox)` or `(id_, checkbox)` or `id_`. MrD didn't like this, referencing the variable length return in a TODO in 3e356302f1f1b82d034e81466b33725026263318. In the case here, it always returns `(None, checkbox)`.
- The `balt.askStyled()` function returns `(None, False) in (wx.ID_OK, wx.ID_YES)`, which is `False`.
- There's no call to actually show `balt.vistaDialog()`. There is a call to `windows.TaskDialog.show()`, but this method returns `tuple(button, radio, checkbox)` and doesn't seem to actually _show_ anything. Conversely, when `canVista==False`, there's actually a call to wx's `ShowModal()`. But maybe I don't understand the Windows API here.
- That said, `button` and `radio` are integers (`ctypes.c_long`), so `balt.vistaDialog()` will never return  `(id_, checkbox)` or `id_`:

```python
    for id_, title in buttons:
        if title.startswith(u'+'): title = title[1:]
        if title == button: # title is a string, button is an integer - these will never be equal
            if checkBoxTxt:
                return id_,checkbox
            else:
                return id_
    return None, checkbox
```

It would be helpful to have some unit tests and more type checks (maybe using `mypy` annotations?).

> Not sure I want to drop it completely - will have to look into it

IMO, relying on the Windows API introduces upstream risk. Perhaps the reason why `balt.vistaDialog()` doesn't work anymore is because of Windows updates? It's equally likely and hard to know for sure. If we're moving to wx3, maybe we should fully commit?

**Comment by fireundubh** [comment]
@Utumno I was waiting to hear back before pushing my changes to my repo, but I didn't so here you go: https://github.com/fireundubh/wrye-bash/tree/fireundubh-wx3-translations-patch

**Comment by Utumno** [comment]
Thanks @fireundubh it's always easier for me to review code and I was pressed for time

> so balt.vistaDialog() will never return (id_, checkbox) or id_:

Thanks - not sure if id_ was ever used at the code (MrD is me)

However, isVista was working fine on wx2 so we need some debugging there before we bin it to make sure it is not possible to fail

Will fetch your branch here and review ASAP - I am curious actually

**Comment by Utumno** [comment]
@fireundubh - what are the functional changes introduced by your commit? It's a bit hard to follow through due to refactoring to class. Is a class and atop level module really needed here? Some remarks:

- new code should be PEP8 apart from things noted in coding style wiki https://github.com/wrye-bash/wrye-bash/wiki/%5Bdev%5D-Coding-Style. Definitelly module names in lower case, attributes in lower_underscore etc
- is a new class needed? Should be new style class
- is a new top level module needed?
- why do you use properties here?
- why are the dicts moved to constants? Are they used in multiple places?

In all, except if there is a reason for the above, it would be better to refactor the initTranslations function. Note that this class makes it harder to follow the logic.

Will have a look at the Vista thing

**Comment by fireundubh** [comment]
> * new code should be PEP8 apart from things noted in coding style wiki https://github.com/wrye-bash/wrye-bash/wiki/%5Bdev%5D-Coding-Style. Definitelly module names in lower case, attributes in lower_underscore etc

I was maintaining the same formatting as the rest of the code. Not my preference. Perhaps I misunderstood you when you said "please do not add renames and formatting differences in your branches"?

> * why are the dicts moved to constants? Are they used in multiple places?

Ideally, they shouldn't be in code; otherwise, having them in a central place makes them more maintainable.

> * is a new class needed? Should be new style class

You'll have to explain your opposition to a new class.

> * is a new top level module needed?

Is this a question about the location of the module, or about the existence of the module? I don't have a preference for the location.

> * why do you use properties here?

Unnecessary accessibility. They could easily be renamed `get*()` because there's no setter.

> In all, except if there is a reason for the above, it would be better to refactor the initTranslations function. Note that this class makes it harder to follow the logic.

You'll have to explain why it's hard to follow. I don't share that view.

**Comment by Utumno** [comment]
On Sun, Jan 13, 2019 at 10:05 AM fireundubh <notifications@github.com>
wrote:

>
>    - new code should be PEP8 apart from things noted in coding style wiki
>    https://github.com/wrye-bash/wrye-bash/wiki/%5Bdev%5D-Coding-Style.
>    Definitelly module names in lower case, attributes in lower_underscore etc
>
> I was maintaining the same formatting as the rest of the code. Not my
> preference. Perhaps I misunderstood you when you said "please do not add
> renames and formatting differences in your branches"?
>
No you didn't - but _new code_ should be PEP8 - so if editing the code it
would be easier for me to not have formating differences so I can readily
see what _really_ changed as opposed to renamed. But if you write a new
class - this should be PEP8 (but see the style guide for differences from
PEP8)

>
>    - why are the dicts moved to constants? Are they used in multiple
>    places?
>
> Ideally, they shouldn't be in code; otherwise, having them in a central
> place makes them more maintainable.
>
Not really if they are used in one place - constants was for big chunks of
data or data imported in many places - just a step when I was splitting
basher

>
>    - is a new class needed? Should be new style class
>
> You'll have to explain your opposition to a new class.
>
1. review what changed instead of having the functionality rewritten with
the changes (somewhere) in it
2. less pythonic - this is really a function. Does it have state? Do you
use the instance? Couple rules of thumb to decide if you need a function.
These things are gut feeling also - are you usually programming in python?
Has a different logic than, say, Java

>
>    - is a new top level module needed?
>
> Is this a question about the location of the module, or about the
> existence of the module? I don't have a preference for the location.
>
Existence - see above

>
>    - why do you use properties here?
>
> Unnecessary accessibility. They could easily be renamed get*() because
> there's no setter.
>
Again not Java - if used once no function please - this is unecessary
complication. I explain: I will have to jump to the function and search
(will do that) where else is used just to find out nowhere and then
continue reading the logic. Complicated - try `import this` on the
interpreter

> In all, except if there is a reason for the above, it would be better to
> refactor the initTranslations function. Note that this class makes it
> harder to follow the logic.
>
> You'll have to explain why it's hard to follow. I don't share that view.
>
Well cause a function is a streamlined thing while several functions I have
to jump back and forth  - but most importantly because that way I can see
what changes are needed in the translation logic I am familiar with,
instead of reading a new piece of code and not really be sure what changed
unless I do the mapping myself.

This is important not only because I need one quarter of the time to review
but because changes in a biggish codebase propagate in ways only people
really familiar with it can appreciate - so I need to see the exact changes.

Would you mind squashing the commits and moving the logic back to the
initTranslation - I can do that and have you review it alternativelly. One
refactoring == 1000 words - but I would rather concentrate on the esl pain

**Comment by fireundubh** [comment]
> if used once no function please - this is unecessary complication. I explain: I will have to jump to the function and search (will do that) where else is used just to find out nowhere and then continue reading the logic. 

That "unnecessary" complication is much easier to test and debug. Jumping to functions is not a big deal. 🤔 

> are you usually programming in python?

Really?

> I would rather concentrate on the esl pain

If you want to solve the locale crash on start, all you need to do is not use `setlocale()` and you're done.

If you don't want to modernize, consider future needs, or write testable code, you don't need what I put up then.

I'm kind of annoyed by your attitude here, so I don't think I'll keep working on this. Good luck.

**Comment by Utumno** [comment]
On Sun, Jan 13, 2019 at 11:39 AM fireundubh <notifications@github.com>
wrote:

> if used once no function please - this is unecessary complication. I
> explain: I will have to jump to the function and search (will do that)
> where else is used just to find out nowhere and then continue reading the
> logic.
>
> That "unnecessary" complication is much easier to test and debug. Jumping
> to functions is not a big deal. 🤔
>
> are you usually programming in python?
>
> Really?
>
What's wrong with asking about your background ?

> I would rather concentrate on the esl pain
>
> If you want to solve the locale crash on start, all you need to do is not
> use setlocale() and you're done.
>
> If you don't want to modernize, consider future needs, or write testable
> code, you don't need what I put up then.
>
It's not a matter of modernizing, I did explain why I do not agree with a
class here. Future needs ? As for tests I am all for that but those need be
written by someone

> I'm kind of annoyed by your attitude here, so I don't think I'll keep
> working on this. Good luck.
>
For posterity I do not think that what I wrote above is attitude, I would
call it a "code review", detailed at that, but let us be judged  by people
reading our posts in the future

**Comment by Infernio** [comment]
An interesting side effect of upgrading to wx3 is that refreshes with many loose files and / or projects are now much faster.

Arthmoor reported a massive speedup, going from around 20 minutes to around **90 seconds**. However, I couldn't reproduce a similar boost with my setup; I only gained around 11 seconds for a full refresh. After some discussion on Discord and looking through `balt.Progress`, I realized the reason for this: updates to the progress bar were expensive in wx2.8, and they were being called so frequently that the rest of the program slowed down as a result.

Specifically, this line in `Progress._do_progress` was responsible:
https://github.com/wrye-bash/wrye-bash/blob/6ff2eb61097ded9e33d78bbfad6031e8dc64a47f/Mopy/bash/balt.py#L1362
The `message != self.prevMessage` comparison would cause a GUI update every time that the message shown on the progress bar changes, which will amount to thousands of times on a full refresh with many loose files and / or projects. This also explains why I couldn't reproduce the boost Arthmoor experienced: my installation has almost no loose files and no projects whatsoever, only ~200 BSAs and 7z archives.

The reason that wx3 sped this up so much is that `wx.ProgressDialog` is now a native Windows widget and GUI refreshes have become very cheap as a result. Taking the `message != self.prevMessage` out gave a speedup of  **1:35** (10:21 -> 8:46) on wx2.8 on a test setup with several thousand loose files, but only **~5 seconds** (8:41 -> 8:35) on wx3 with the same setup.
Don't know if it's really worth removing the 'update every time the text changes' behavior for a 0.097% speedup.

**Comment by Utumno** [comment]
Well well - I had the same impression (I had many loose files in TESIV setup) and I kinda knew the reason (I have explicitly commented on that but was still not sure it was just an impression of mine, https://stackoverflow.com/questions/42890766/wx-python-3-0-2-classic-disable-the-cancel-button-of-wx-progressdialog#comment73005697_42890766 - Robin Dunn is the wx maintainer).
Thanks for confirming and detailed timings - green light from me for speedups and/or refactoring, the Progress class is one of the most ancient in here and does need some dusting (it's not clear how to use sometimes and its SubProgress derived one, plus it is not properly destroyed, will push some half baked fixes)

----

Gosh, I spent so much effort on those wx3 things ( https://stackoverflow.com/search?q=user%3A281545+is%3Aquestion+%5Bwxpython%5D ), glad you turned the Eye on them. We can spend some time together to make sure that all half baked fixes from utumno/metallicow/nycz branches are ported.

**Comment by Infernio** [comment]
I think the only remaining broken thing in the nightly is the Cancel button on the installers tab refresh, because I lost the half-fix commit you made for that :P

**Comment by Utumno** [comment]
It's a half fix however: https://github.com/wrye-bash/wrye-bash/commit/aa89829851550aa384bf03fbd623fc573ea89eae

**Comment by Infernio** [comment]
Just found a really nasty crash: Opening, closing and then reopening the mod checker hard-crashes python. Doesn't happen with doc browser and WryeLog-based things.

**Comment by Infernio** [comment]
Oh, found it :P
It's even a TODO:
https://github.com/wrye-bash/wrye-bash/blob/7aef1e1576ac76fd3a06ba8f81b6dee7a2dac1f5/Mopy/bash/balt.py#L3315-L3319

**Comment by Utumno** [comment]
Yep was not sure why they kept it alive - probably it was thought costly to
create - just set it to none (btw this is the case for the patcher classes,
they are created once, and this I believe is the root of the huge memory
leak in python patcher)

**Comment by Infernio** [comment]
The workaround for the wx.Locale stuff (https://github.com/wrye-bash/wrye-bash/commit/dab900c3b804d392a7f57d7481766ed815ac5acc#diff-b7b74b62524654cab1e196feda059de1R312-R319) doesn't seem to cover everyone. Someone on Discord is reporting that they still get the crash, and it's preventing the creation of a BashBugDump as well, so I can't even tell what their locale is (not likely that they'd know either). My only guess is that the locale is an uncommon one that is supported by wx3 but not py2, so a discrepancy between setlocale and wx.Locale(wx.LANGUAGE_DEFAULT) is inevitable. Not even sure if we *can* work around that...

**Comment by Utumno** [comment]
Is it a hard crash? we did develop a mechanism in bash.py to catch such
crashes - maybe add the bugdumb to the message displayed there?

**Comment by Infernio** [comment]
Looks like Progress is causing even more issues. If you're on the Mods tab, then minimize WB, change something in the Data folder that causes a Mods refresh (e.g. deleting a plugin) and then click on the WB icon in the taskbar, it will briefly open, show a progress bar, but collapse instantly and flash orange in the task bar. Clicking the icon again after that opens WB normally.
Seems like Progress maybe steals the focus, then dies and the focus is not transferred to the main window?
Wasn't able to reproduce with the other tabs though, so it might not be a universal `balt.Progress` issue.

TODO: Confirm if this only occurs on wx3, I think I experienced something like this on wx2.8 too...

**Comment by Infernio** [comment]
One user who still gets the wx.Locale crash sent me a screenshot of their Windows language setting. I'll try reproducing it with that once I have access to a Windows PC again:
![locale_crash](https://user-images.githubusercontent.com/2845240/65823492-a03e9980-e257-11e9-8f6c-eb8062f9e902.png)

**Comment by Infernio** [comment]
According to Discord feedback, 8ce842b7edd68d7c4dd1aa38ef1b5c9005404eca seems to have *finally* put the locale crash to rest.

**Comment by Utumno** [comment]
Agree - I had come near that on my dry runs but again all the format*** methods in bolt failed to do anything useful. Well, we could fill an issue for fixing that come 308 or/and py3 (I am not sure what the situation is in phoenix but I am affraid it's the same). Needless to say, this had cost me quite some time - I had even posted on wx list:

Change locale by creating wxLocale objects in wxpython 3.0.2.0 - Google Groups
https://groups.google.com/forum/#!topic/wxpython-dev/0y14W8Mp2sg

python - Change locale by creating wxLocale objects in wxpython 3.0.2.0 - Stack Overflow
https://stackoverflow.com/questions/54178028/change-locale-by-creating-wxlocale-objects-in-wxpython-3-0-2-0

python - wxpython 3.0 breaks older apps? (locale error) - Stack Overflow
https://stackoverflow.com/questions/21444951/wxpython-3-0-breaks-older-apps-locale-error

Set locale the wxPython way - Google Groups
https://groups.google.com/forum/#!topic/wxpython-users/fDHE9uujaZA

wxPython is unrunnable due to 'locale error' - Google Groups
https://groups.google.com/forum/#!topic/wxpython-users/QxU9FtHrnhox

**Comment by Infernio** [comment]
`locale.format` needs a `locale.setlocale` call to work. However, `setlocale` is what's causing the mismatch. So, we have two choices:
 1. Write *very* careful logic to ensure we always create a `wx.Locale` object that matches the `setlocale` call
 2. Kill `setlocale` and only use `wx.Locale`
I've tried the first option many, many times. Some obscure setup always sneaks its way through - first it was German, then Mexican, then I thought I'd finally gotten it right, when someone with their locale set to English, but their date formatting set to Hebrew showed up and I had to throw it all out again because `getlocale` returned `he_IL` (Hebrew, not English) in that case. At that point, I gave up and killed the `setlocale` call.

**Comment by Utumno** [comment]
Ok let's kill it then - although maybe bugging the wx guys for at least an explanation _why_ they messed up with python's setlocale standard practice could be of interest, duh!

**Comment by Infernio** [comment]
I've set up a branch for an eventual merge of wx3 support here: [15-wx3-upgrade](https://github.com/wrye-bash/wrye-bash/tree/15-wx3-upgrade).

**Comment by Utumno** [comment]
Great - I was thinking of moving some of the changes that are wx2 compatible to a separate commit and pushing that (also some noop commits from nightly if you don't mind, like the load order warns etc - would prefer them squashed if you don't mind, I can do that)

**Comment by Infernio** [comment]
I can `git gui` together a commit that's backwards compatible with wx2.8, yeah. I'll push it to some separate branch for review.

**Comment by Utumno** [comment]
Thanks - this will help us compare with all those branches that are on wx 3 and prune them (rabasing them on this common commit - I can do that)

**Comment by Utumno** [comment]
Left some comments for you at for-review @Infernio - let's carry all wx2 compatible stuff and merge. Another branch that needs deleting is https://github.com/wrye-bash/wrye-bash/tree/metallicow-wx-compatibility-rebased-on-dev but it may have some stuff that we need to use - the first few commits which however after a quick look are already taken care of in for_review (aka 15-wx3-compatibility)

@GandaG please rebase ganda-431-lootganda-260-deployganda-415-virtualenv on current dev ommiting all the nightly content - that way it will be easy to squash cherry-pick on nightly for testing without carrying all the (sometimes outdated) nightly content

Thanks :)

**Comment by Infernio** [comment]
> Another branch that needs deleting is https://github.com/wrye-bash/wrye-bash/tree/metallicow-wx-compatibility-rebased-on-dev but it may have some stuff that we need to use 

I'll have to dig through that one carefully, but the first few commits are definitely unnecessary. Some of the later phoenix stuff has better solutions as well, but we'll see about that when we get there ;)

**Comment by Utumno** [comment]
Deleted it - moved what's useful from it into utumno-wx3 - left a comment on the only commit that is of interest. Do you think there is any useful stuff in metallicow-phoenix-skeleton or metallicow-wx3 ?

**Comment by Infernio** [comment]
metallicow-wx3: Some of those changes are needed for phoenix, but the commit is so old that almost all of its changes have already been made it to dev, so probably fine to delete.
metallicow-phoenix-skeleton: Looks like a POC for implementing Wrye Bash's GUI using [wx.aui](https://wxpython.org/Phoenix/docs/html/wx.aui.1moduleindex.html). Might be interesting to keep it around for GUI inspiration in the future?

**Comment by Metallicow** [comment]
Umm try whatever you want. Beware when you hit wx4.0 you might want to stop, as 4.1 is a breaker or as in the situation you are in now. Working with wxPy myself and helping the progress, I can say it will be less headache for bashers to deal with. 2.8 is quite buggy, but it only gets better from here. It's rocket science.

**Comment by Metallicow** [comment]
I can't give you the skeleton for SoureCoder, but you could build one. The Concept is pretty simple rebuilt the GUI and add the ability to modify it in real time and reload it. The basics is built into python but the words are the key. it is abstract when you do this so have fun or beware. it is necessary to destroy the program in order to modify it. nuff said.

**Comment by Metallicow** [comment]
Most of the GUI work can be done in SourceCoder in less than an hour. Time yourself. Destruction magick imminent.

**Comment by Utumno** [comment]
@Infernio repushed the branch but messed up a bit rebase so let me rerepush - meanwhile I noticed yet another bug

 - when having a warning displayed as in
 
        if message:
            ListBoxes.Display(
              self, _(u'Warnings'), _(u'The following warnings were found:'),
            message, liststyle='list', canCancel=False)

 and we close the window by the main X button python crahes and the windows dialog for stopping/debugging is displayed

- Progress - whenever we close a progress window via Cancel or automatically Wrye loses focus

**Comment by Utumno** [comment]
Ok rerepushed - included my hacky workaround which should land on dev along with patcher fixups (maybe squashed? or non ff merge?)

8eb0e39c82a1f6a52dd5613eb9334812eb98cd38
8b07532e717405199d163e6bfa1bbf3ac316f17b
ac6f7380378ed1a052cec173d5f4519c9c22eb12
6cea5eea6afa56a27e1c6a8e893fbf47f194f04b
0b7b7099d3a08253572844537115b0e7896184a0
cf3bb8bc5e3a5f69d1f89c0f173007d58a6db246
f00aab52d12bd2d6df57c9392f54533c9afbbd45
47ca8d5067e396c55d3137ef706041f53b96f99f

and BAIN ones

4ce8c8c68146f98ea0be06929dc537c7424a2193 + d8446a23e14aea855896489167953ccd5955bf57
ff363ef2999cd391f7cf35b273d564e6999d2076

Squash that in too b962d8dbf61225aae1f5b139d151072ad03636ef

Re: load order - good job! I think that if we implement moving the active lists in BashLoadOrder.dat we can close that issue with a nice non ff merge

Of course I may be missing something and maybe some of it needs more testing

**Comment by Utumno** [comment]
I may have missed smth (or added something stupid) but you get the picture - another annoying TODO: chardet merge should try and download chardet in the python version of the installer then we can merge that too

**Comment by Infernio** [comment]
> python crahes and the windows dialog for stopping/debugging is displayed

Oh god, this sounds like fun. Will spend some time looking into it.

> whenever we close a progress window via Cancel or automatically Wrye loses focus

Yep, that one has been getting a lot of attention on Discord. No idea what causes it yet.
While we're here, another issue that keeps coming up:
 - [ ] wx3 fixed the ugly white widget backgrounds (yay!), but this has had the side effect of exposing how our splitters are completely invisible, making resizing the bash tags box a nightmare.

**Comment by Utumno** [comment]
Thanks - we could maybe add a Link.Frame.Show() when progress is destroyed although this might cause Wrye to pop up while in the background? For another example related check ModChecker - push that button "scan for dirty edits" then cancel the progress - Bash disappears (!)

**Comment by Infernio** [comment]
Decided to do the records/patcher merge as `no-ff`, since the commits are quite diverse and one will have to be reverted later (`FO3/FNV: Disable Import Relations`).

**Comment by Infernio** [comment]
So, the real bug here seems to be that clicking the X of the main window shouldn't even be possible, since `ListBoxes.Display` should open a modal dialog (see here: https://wxpython.org/Phoenix/docs/html/wx.Dialog.html#phoenix-title-modal-and-modeless). For some reason this works fine with `wx.TreeCtrl`, the one we use for the 'masters have been activated / children have been deactivated' warning, but not with `wx.ListBox`.

**Comment by Infernio** [comment]
I think I know why it happens, but this might not be fixable. wx makes its modal dialogs work by creating a temporary event loop and pausing the application's main loop. However, these modal dialogs happen before we've created our main loop (`app.MainLoop()` in bash), so the temporary event loop probably just keeps everything that doesn't have an event loop alive, including the (not yet fully initialized) GUI. Maybe we need to move `RefreshData` to [OnEventLoopEnter](https://wxpython.org/Phoenix/docs/html/wx.AppConsole.html#wx.AppConsole.OnEventLoopEnter)?

**Comment by Utumno** [comment]
If it's fixable on wx4 then I'd say we can live with it as no user is uspposed to open Bash just to peek those warnings. Good you investigate wx boot however - fixing the weird focus issues may be also related to all this

**Comment by Metallicow** [comment]
If you wrap every call if anything has a widget, it will help, tho testing old code and new will result in a test havok if tests are made. Ex: dont write unittests for 2.8 and expect them to work for 4.0 without full attention for example. If you can wrap all wx4 code before you get there... even better, tho all logic code is your own problem.

**Comment by Infernio** [comment]
Interesting factoid: the 'progress dialog causes WB to lose focus' bug doesn't seem to happen with pyinstaller 🤔

**Comment by Metallicow** [comment]
@Infernio Most widget focus bugs are worked out in wx4.0, tho they might still appear in 2.8 <> 4.0.
Otherwise it is an internal problem.

**Comment by Utumno** [comment]
I suspect that the loose focus bug has to do with saving window size/position - whenever I experience it on boot the warnings windows are messed up (minimum size ?) - if the bug does not kick in they display normally - so related to the code that saves position for not maximized windwos

Just wild guesses

**Comment by Infernio** [comment]
For reference's sake: the locale situation seems to maybe have finally gotten better in wx4.1: https://github.com/wxWidgets/Phoenix/pull/1702
